google-cloud-spanner/src/main/java/com/google/cloud/spanner/TransformingOperationFuture.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2017 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.spanner;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.api.core.ApiFuture;
+import com.google.api.core.ApiFunction;
+import com.google.api.gax.longrunning.OperationFuture;
+import com.google.api.gax.longrunning.OperationSnapshot;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+/** A wrapper class to augment proto results returned from get() */
+public class TransformingOperationFuture<ResponseT, RawResponseT, MetadataT>
+    implements OperationFuture<ResponseT, MetadataT> {
+
+  private OperationFuture<RawResponseT, MetadataT> rawOperationFuture;
+  private ApiFunction<RawResponseT, ResponseT> transformingFunction;
+
+  public TransformingOperationFuture(
+      OperationFuture<RawResponseT, MetadataT> rawOperationFuture,
+      ApiFunction<RawResponseT, ResponseT> transformingFunction) {
```

Comment:
I think we should add exception transformer here too. That way we can transform the exception from `get` into `SpannerException`, and callers don't need to change.

Reply:
@garrettjonesgoogle @pongad Sorry if I didn't catch the entire idea of this. Even if we can convert any of `ExecutionException`, `TimeoutException` or `InterruptedException` into a `SpannerException`, as long as we can't change the signature of `get()` and `get(long, Unit)` (since `TransformingOperationFuture` implements `OperationFuture`), callers still have to catch these exceptions, though they don't have to do the conversion themselves. Any ideas to work around this?

Reply:
We should get buy in from others first but I think this is fine. We are changing from `Operation` type to `Future` anyway. My thinking is that `get()` should throw `ExecutionException` but the cause of the exception should be a `SpannerException`.